public/app/features/alerting/unified/components/rules/state-history/LokiStateHistory.tsx

```diff
@@ -0,0 +1,234 @@
+import { css } from '@emotion/css';
+import { isEmpty, sortBy, take, uniq } from 'lodash';
+import React, { useCallback, useMemo, useRef, useState } from 'react';
+import { useForm } from 'react-hook-form';
+
+import { dateTime, GrafanaTheme2, TimeRange } from '@grafana/data';
+import { Stack } from '@grafana/experimental';
+import { Alert, Button, Field, Icon, Input, Label, TagList, Tooltip, useStyles2 } from '@grafana/ui';
+
+import { stateHistoryApi } from '../../../api/stateHistoryApi';
+import { combineMatcherStrings } from '../../../utils/alertmanager';
+import { HoverCard } from '../../HoverCard';
+
+import { LogRecordViewerByTimestamp } from './LogRecordViewer';
+import { LogTimelineViewer } from './LogTimelineViewer';
+import { useRuleHistoryRecords } from './useRuleHistoryRecords';
+
+interface Props {
+  ruleUID: string;
+}
+
+const MAX_TIMELINE_SERIES = 12;
+
+const LokiStateHistory = ({ ruleUID }: Props) => {
+  const styles = useStyles2(getStyles);
+  const [instancesFilter, setInstancesFilter] = useState('');
+  const logsRef = useRef<HTMLDivElement[]>([]);
+
+  const { getValues, setValue, register, handleSubmit } = useForm({ defaultValues: { query: '' } });
+
+  const { useGetRuleHistoryQuery } = stateHistoryApi;
+  const timeRange = useMemo(() => getDefaultTimeRange(), []);
+  const {
+    currentData: stateHistory,
+    isLoading,
+    isError,
+    error,
+  } = useGetRuleHistoryQuery({ ruleUid: ruleUID, from: timeRange.from.unix(), to: timeRange.to.unix() });
+
+  const { dataFrames, historyRecords, commonLabels, totalRecordsCount } = useRuleHistoryRecords(
+    stateHistory,
+    instancesFilter
+  );
+
+  const frameSubset = useMemo(() => take(dataFrames, MAX_TIMELINE_SERIES), [dataFrames]);
+
+  const onLogRecordLabelClick = useCallback(
+    (label: string) => {
+      const matcherString = combineMatcherStrings(getValues('query'), label);
+      setInstancesFilter(matcherString);
+      setValue('query', matcherString);
+    },
+    [setInstancesFilter, setValue, getValues]
+  );
+
+  const onFilterCleared = useCallback(() => {
+    setInstancesFilter('');
+    setValue('query', '');
+  }, [setInstancesFilter, setValue]);
+
+  const onTimelinePointerMove = useCallback(
+    (seriesIdx: number, pointIdx: number) => {
+      const uniqueTimestamps = sortBy(uniq(frameSubset.flatMap((frame) => frame.fields[0].values)));
```

Comment:
Similar concerns to the ones I had above – can we abstract this in to a separate function that takes `(frames: DataFrame[], index: number)`? We can maybe add memoization here if we think we need it later to optimize for perf.